### ISSUE TITLE
for loop edge case freezes yaegi

### ISSUE THREAD

**Issue by nosarthur** [containous/yaegi#340]
The following program `sample.go` triggers a panic:

```go
package main

import "fmt"

func main() {
        v := []int{1, 2, 3}
        for i := range v {
                v = append(v, i)
        }
        fmt.Println(v)
}
}
```

Expected result:
```console
$ go run ./sample.go
// ouput
[1 2 3 0 1 2]
```

Got:
```console
$ yaegi ./sample.go
// ouput
```
There is no output, it seems to get stuck

**Comment by kortschak** [comment]
The issue here is, in part, that value recovery by `genValueArray` always goes back to the node's frame. In the case of a `range` operation this is not correct since a `range` initially copies the value of the range expression before the loop is executed; "The range expression x is evaluated once before beginning the loop, with one exception: if at most one iteration variable is present and len(x) is constant, the range expression is not evaluated."

This is demonstrated by the following code that also fails to execute correctly.
```
package main

import "fmt"

func main() {
	a := [...]int{2, 1, 0}
	for _, v := range a {
		a[v] = v
	}
	fmt.Println(a)
}

// Output:
// [0 1 2]
```
```
$ yaegi range1.go 
[2 1 2]
$ go run range1.go 
[0 1 2]
```

Confirming this, the range on a `*[n]int` does reflect the compiled case as expected since the range is on the pointer, not the array.

```
package main

import "fmt"

func main() {
	a := [...]int{2, 1, 0}
	for _, v := range &a {
		a[v] = v
	}
	fmt.Println(a)
}

// Output:
// [2 1 2]
```

```
$ yaegi range2.go 
[2 1 2]
$ go run range2.go 
[2 1 2]
```

**Comment by kortschak** [comment]
I have found issues with the map range handling as well.

Of the following three cases, the first and last fail with a panic. I have a change that fixes all of these. However it depends on `reflect.MapIter` which was introduced in Go1.12. Currently you test back to Go1.11, but only claim support for Go1.12 onward.

```
package main

import "fmt"

func main() {
	m := map[int]bool{1: true, 3: true, 5: true}
	for k := range m {
		m[k*2] = true
	}
	fmt.Println("ok")
}

// Output:
// ok
```

```
package main

import "fmt"

func main() {
	m := map[int]bool{1: true, 3: true, 5: true}
	for _, v := range m {
		fmt.Println(v)
	}
}

// Output:
// true
// true
// true
```

```
package main

import "fmt"

func main() {
	m := map[int]bool{1: true, 3: true, 5: true}
	var n int
	for range m {
		n++
	}
	fmt.Println(n)
}

// Output:
// 3
```

```
package main

import (
	"fmt"
	"math"
)

func main() {
	m := map[float64]bool{math.NaN(): true, math.NaN(): true, math.NaN(): true}
	for _, v := range m {
		fmt.Println(v)
	}
}

// Output:
// true
// true
// true
```

**Comment by mvertes** [comment]
Fixed by #392 and #393
